File: adt2unit/get_chanpos.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

uint8_t get_chanpos (const void *data, uint8_t channels, uint8_t scancode)
{
  uint8_t i = 0;

  do
  {
    const uint8_t *buf = (const uint8_t *)data + i;
    uint8_t n = channels - i;
    bool f = !n;

    while (n)
    {
      f = scancode == *buf;
      buf++;
      n--;
      if (!f)
        break;
    }

    if (!f)
      return 0;

    if (!is_4op_chan (channels - n))
      return channels - n;

    i++;
  }
  while (true);
}

File: adt2unit/i386/get_chanpos.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

function get_chanpos(var data; channels,scancode: Byte): Byte;

var
  result: Byte;

begin
  asm
        xor     ebx,ebx
@@1:    mov     edi,[data]
        add     edi,ebx
        xor     ecx,ecx
        mov     cl,channels
        mov     al,scancode
        sub     ecx,ebx
        repnz   scasb
        jnz     @@2
        xor     eax,eax
        mov     al,channels
        sub     eax,ecx
        jmp     @@3
@@2:    xor     eax,eax
        jmp     @@5
@@3:    pusha
        push    eax
        call    is_4op_chan
        or      al,al
        jz      @@4
        popa
        inc     ebx
        jmp     @@1
@@4:    popa
@@5:    mov     result,al
  end;
  get_chanpos := result;
end;
